### PR TITLE
Support ANY type in Header functions And Planning of Control Flow Functions. 

### PR DESCRIPTION
**Pull request by yliuuuu**
## Relevant Issues
- N/A

## Description
- Control Flow functions such as `coalesce` and `nullIf` will be rewrited to a case when expression. 
- Support for ANY type in Function Header. 

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
  - No, feature branch. 

- Any backward-incompatible changes? **[YES/NO]**
  -  No. 

- Any new external dependencies? **[YES/NO]**
  -  No.

- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**
Yes. 

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt [hunk 1]
```diff
@@ -326,8 +337,29 @@ object PartiQLHeader : Header() {
         )
     }
 
-    // TODO
-    private fun coalesce(): List<FunctionSignature.Scalar> = emptyList()
+    // COALESCE(expression, expression, ... )
+    // Initial implementation of Coalesce.
+    // As the number of parameter can not be pre-determined, we wrap those into a list
+    private fun coalesce(): List<FunctionSignature.Scalar> = listOf(
```
Comment (yliuuuu) [inline partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt:343]: Another way to model the control flow functions such as `coalesce`/`nullIf` is to rewrite them using `Case When` during planning. That is in my opinion a better approach.

File: partiql-lang/build.gradle.kts [hunk 2]
```diff
@@ -85,5 +85,6 @@ tasks.processTestResources {
 
 tasks.shadowJar {
     archiveBaseName.set("shadow")
+    exclude("**/*.kotlin_metadata")
```
Comment (yliuuuu) [inline partiql-lang/build.gradle.kts:88]: https://youtrack.jetbrains.com/issue/KT-25709/IDE-Unresolved-reference-from-fat-jar-dependency-with-Kotlin-runtime

File: partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt (hunk 1 above)
Comment (RCHowell) [reply partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt:343]: Let's mark that as a TODO for the next milestone. This works well enough for us now.

**Review by RCHowell — CHANGES_REQUESTED** [review]

File: partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt (hunk 1 above)
Comment (yliuuuu) [reply partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt]: After some consideration, I believe that it is better to have the rewrite implemented now.

File: partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt [hunk 3]
```diff
@@ -326,23 +384,6 @@ object PartiQLHeader : Header() {
...
-            returns = t,
-            parameters = listOf(
-                FunctionParameter("value", t),
-                FunctionParameter("nullifier", BOOL), // TODO: why is this BOOL?
-            ),
-            isNullCall = true,
-            isNullable = true,
-        )
-    }
-
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt:345]: goodbye!!

File: partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt [hunk 4]
```diff
@@ -484,14 +632,26 @@ object PartiQLHeader : Header() {
                 )
                 operators.add(signature)
             }
+            val anySignature = FunctionSignature.Scalar(
+                name = "date_diff_${field.name.lowercase()}",
+                returns = ANY,
+                parameters = listOf(
+                    FunctionParameter("interval", INT),
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt:639]: Should interval also be ANY?

**Review by RCHowell — APPROVED** [review]

File: partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt (hunk 4 above)
Comment (yliuuuu) [reply partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLHeader.kt]: good catch : )

**Pull request merged**